File: iosApp/iosApp/Functions/DSL.swift
import Foundation
import shared

class DslExample{
    private func dslBlock(block: (Dsl) -> Dsl) -> Dsl {
        var dsl = Dsl()
        dsl = block(dsl)
        return dsl
    }
    
    func example(){
        Dsl().experiments { experimentsDsl in
            experimentsDsl.enable(experiment: Experiment(key: "key1", description: "desc1"))
        }
    }
}

File: kotlin-swift-interopedia-samples-without-skie/shared/src/commonMain/kotlin/com/jetbrains/swiftinteropplayground/functions/DSL.kt
package com.jetbrains.swiftinteropplayground.functions

@DslMarker
annotation class DslMarkerExample

data class Experiment(val key: String, val description: String)

@DslMarkerExample
class ExperimentsDsl() {
    val experimentList: MutableList<Experiment> = mutableListOf<Experiment>()

    fun enable(experiment: Experiment) {
        experimentList += experiment
    }

}

@DslMarkerExample
class Dsl() {
    val experimentsDsl: ExperimentsDsl = ExperimentsDsl()

    fun experiments(block: ExperimentsDsl.() -> Unit = {}) {
        experimentsDsl.block()
    }
}

fun example() {
    val dsl = Dsl().apply {
        experiments {
            enable(Experiment(key = "key1", description = "desc1"))
        }
    }
}